### DISCUSSION TITLE
Lazy Initialization of Cosmos Client: First call taking significant amount of time ~20 seconds

### DISCUSSION THREAD

**Discussion by creedm3** [IEvangelist/azure-cosmos-dotnet-repository#305] [Q&A]
Hey all,

I tried to find this in the discussion, but wasn't exactly sure what term to search for (searched for lazy and init)...

I recently started using this project (love the in memory testing feature, this needs to be a standard for all projects!) and am using a standard implementation of an IRepository.

The behavior I am noticing is on the first call it is taking a significant amount of time to query cosmos... Is this by design? Subsequent calls I am seeing much more familiar times. 

Was wondering if there is a config or health check I am missing or if this is standard...

**Comment by IEvangelist** [comment]
Wow, yeah 20 seconds seems really long. That shouldn't be the case. We are using all sorts of lazy initialization, the idea being that during app startup - it will take a bit to get out of the "cold" state, but once you're moving it shouldn't be a hindrance to performance. I'd be curious if you've done an benchmarks, side-by-side of the Cosmos DB .NET SDK, and our wrapper. That's actually a good idea, perhaps we should do that. What do you think @mumby0168?

**Reply by mumby0168** [reply]
Yeah it would depend on a lot on your configuration, we offer a lot of sync capabilities and some can take longer than others, would it be possible for you to share your RepositoryOptions configuration, along with any container config you have implemented @creedm3 ?

**Reply by creedm3** [reply]
I am really just using the default configs...

                services.AddCosmosRepository(
                    options =>
                    {
                        options.CosmosConnectionString = secretClient.GetSecret("CosmosDbConnectionString").Value.Value;
                        options.ContainerId = Configuration["CosmosDb:JobsContainerName"];
                        options.DatabaseId = Configuration["CosmosDb:DatabaseName"];
                    },
                    options =>
                    {
                        options.SerializerOptions = new CosmosSerializationOptions
                            { PropertyNamingPolicy = CosmosPropertyNamingPolicy.CamelCase };
                        options.HttpClientFactory = () =>
                            services.BuildServiceProvider()?.GetService<IHttpClientFactory>()!.CreateClient(cosmosDbHttpClientName);
                    });

and then not really doing anything fancy other than injecting the IRepository<MyType> into my controller class.

**Reply by creedm3** [reply]
maybe adding the health check would help because it would get some of that initialization out of the way?
